### PR TITLE
BFD-2728: Create notifications for any 500s

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2728](https://jira.cms.gov/browse/BFD-2728)

**User Story or Bug Summary:**

We want to improve observability within our `prod-sbx` (and other) environment(s) so that we are alerted to issues that may be occuring there.

---

#### Diagram

image

#### Overview

This PR implements Slack alerts for _any_ `500`s (errors) logged by BFD Server. Whenever a particular environment's `server` logs a `500`, an alert will be sent within 10 to 15 seconds, and then additional alerts will be sent periodically approximately every 5 minutes until the errors cease in the given environment.

As it is possible that _many_ errors occur in a short period, these alerts summarize the number of errors that have been returned during a "lookback" period (330 seconds to account for variance in time between recurring invocations of the alert Lambda). These summaries include a count of `500`s per-partner, per-API-operation, along with a total number of errors returned within the "lookback" period.

Along with the summary, the alert includes various links to Log Insights queries that allow an operator to view the log messages related to the errors with more detailed information such as exception stack traces, request URLs, etc.

To reduce cost, the "scheduler" Lambda acts as an itermediary between the CloudWatch Alarm alerting on _any_ `500`s and the "alerter" Lambda that queries the CloudWatch Logs. As scanning logs can be expensive, the "alerter" Lambda is only ever running if the `500`s CloudWatch Alarm is in `ALARM`.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [ ] Yes
    - [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- N/A

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by mjburling — COMMENTED** [review]
Curious if you'd be opposed to storing the source and output of the diagram that describes the module with the code base.

**Comment by malessi** [comment]
> Curious if you'd be opposed to storing the source and output of the diagram that describes the module with the code base.

Not opposed at all. I put together the diagram for personal use, hence why it's a less _polished_ than our typical diagrams, but I think it's useful enough as-is to store in the repository alongside this new module.

This reminds that I forgot to include a `README.md` with this new module, so I'll do that and add the diagram there. I'm not sure exactly where I want to keep the source code for the diagram itself, though. Perhaps just a `docs/diagram` subfolder relative to the module root? I'll figure it out.

File: ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            )
+            scheduler_client.delete_schedule(
+                GroupName=EVENTBRIDGE_SCHEDULES_GROUP, Name=schedule.name
+            )
+            logger.info("Schedule %s deleted successfully", schedule.name)
+    except scheduler_client.exceptions.ClientError:
+        logger.error(
+            "An error occurred when trying to delete old OnetimeSchedules; continuing. Err: ",
+            exc_info=True,
+        )
```
Comment (malessi) [inline ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter_scheduler.py:192]: One-time EventBridge Schedules [do not automatically remove themselves](https://docs.aws.amazon.com/scheduler/latest/UserGuide/managing-schedule-delete.html) after executing, so before doing any work the Scheduler will remove expired one-time schedules from the corresponding Schedule Group. This way, at most we'll only ever have a single, expired one-time schedule in each environment's Schedule Group.

Comment (mjburling) [reply ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter_scheduler.py:192]: Thanks for the explanation!

File: ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+| filter ispresent('exception')
+| filter ispresent('mdc.http_access_request_uri')
+| filter level = 'ERROR'
+| sort @timestamp desc
+""".strip()
+LOG_INSIGHTS_MESSAGES_JSON_CROSS_REF_QUERY = """
+fields @message,
+`mdc.http_access_request_header_Bluebutton-Originalquerytimestamp` as orig_timestamp
+| filter orig_timestamp == 'REPLACE WITH MDC ORIGINAL TIMESTAMP OF MESSAGE FROM ACCESS.JSON'
+""".strip()
```
Comment (malessi) [inline ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py:56]: Is this query useful to include as part of the query links in the Slack notification? Log Insights has [no function for joining fields](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL_QuerySyntax.html) from two log sources (say, `messages.json` and `access.json`), so crafting a query to do this "cross-referencing" automatically is impossible. I wanted to include a _somewhat_ convenient way for an engineer responding to an alert to be able to see exception details for a particular `500` and this half-solution was all I could come up with.

The problem here is that each Log Insights URL added to the Slack message alert balloons its size considerably, as the query needs to be embedded within the URL that's posted with the alert. Slack messages have a [limit of about 4000 characters](https://api.slack.com/docs/rate-limits#rtm-posting-messages) and with this additional Log Insights query link the message is getting close to the limit; for example, an alert with a single partner table with only a single row is already 3636 characters long. Removing the link to this particular "cross-referencing" query reduces that to _2782 characters_, a difference of 854 characters.

Does anyone feel strongly that including a link to this query (that will query the time period associated with the alert) is useful enough to keep?

File: ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        table_block_str = (
+            f"```\n{'Operation':<{longest_op_chars_count}} {'Error Count':<{longest_err_chars_count}}\n"
+        )
+        for op, count in op_counts.items():
+            table_block_str += (
+                f"{op:<{longest_op_chars_count}} {count:<{longest_err_chars_count}}\n"
+            )
+        table_block_str += "```"
+
+        per_partner_tables[partner] = table_block_str
```
Comment (malessi) [inline ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py:241]: Slack webhooks [do not seem to support Markdown-formatted tables](https://stackoverflow.com/questions/37080635/slack-webhook-html-table), and using the rudimentary [`fields`](https://stackoverflow.com/a/55734396) section for display a two-column table is limited to 10 entries.

So, this code generates an appropriately formatted "console table", using code fences to display unformatted, monospaced text. For example:

```
Operation                  Error Count
this/is/a/long/operation() 99999999
short/operation()          10000000
...
```

File: ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py (hunk 2 above)
Comment (malessi) [reply ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py:56]: Yeah, I'm thinking this particular query is probably more suited for inclusion in a runbook (which I also was considering writing but this story has already taken long enough as it is). However, I _do_ think the other two queries are useful as I suspect an interested engineer will want to quickly know _why_ we're returning errors and what the requests that are causing them look like. With the current alert, they could click one of the links and immediately be able to start investigating what's going on. With just a link to a runbook, there's just a _little_ extra friction there that I feel could lead to a less rapid response.

At the bare minimum, I'm going to remove `LOG_INSIGHTS_MESSAGES_JSON_CROSS_REF_QUERY` and all usages of it, including the link in the Slack message. I would like to keep the links to the other two queries, and maybe in the future, when the corresponding runbook exists, we can add a third link to it as well.

File: ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    )
+    slack_message = {
+        "blocks": [
+            {
+                "type": "section",
+                "text": {
+                    "type": "mrkdwn",
+                    "text": (
+                        ":notification-alert: *Error Alert for"
+                        f" `{BFD_ENVIRONMENT}` BFD Server* :notification-alert:\n"
```
[the remark this answers is not in the document]
Comment (mjburling) [reply ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py:258]: The disco notification is fun.

[the remark this answers is not in the document]
Comment (malessi) [reply ops/terraform/services/server/modules/bfd_server_error_alerts/lambda_src/bfd_server_errors_alerter.py]: After the a11y concerns brought up during the recent demo, I've changed the emoji to `:alert:` (⚠️).

**Review by mjburling — APPROVED** [review]

**Review by JeremyIR — APPROVED** [review]

**Review by aadedejifearless — APPROVED** [review]

**Pull request merged**